hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/TssBaseServiceImpl.java

```diff
@@ -380,12 +391,255 @@ public TssMessage getTssMessageFromBytes(Bytes wrap, TssParticipantDirectory dir
+        // collect tss encryption keys for all nodes in the active roster that are not null
+        final var encryptionKeys = activeRoster.rosterEntries().stream()
+                .map(entry -> tssStore.getTssEncryptionKey(entry.nodeId()))
+                .filter(Objects::nonNull)
+                .toList();
+
+        final var info = new RosterAndTssInfo(
+                activeRoster,
+                requireNonNull(rosterStore.getCurrentRosterHash()),
+                rosterStore.getCandidateRoster(),
+                targetRoster,
+                tssStore.getMessagesForTarget(targetRoster),
+                tssStore.anyWinningVoteFrom(rosterStore.getCurrentRosterHash(), targetRoster, rosterStore),
+                encryptionKeys,
+                selfVote);
+        manageTssStatus(info, isStakePeriodBoundary, consensusNow);
+    }
+
+    void manageTssStatus(final RosterAndTssInfo info, final boolean isStakePeriodBoundary, final Instant consensusNow) {
+        final var statusChange = new StatusChange(isStakePeriodBoundary, consensusNow, info);
+        final var newTssStatus = statusChange.computeNewStatus();
+        if (!newTssStatus.equals(tssStatus)) {
+            tssStatus = newTssStatus;
+        }
     }
 
     @VisibleForTesting
     public TssKeysAccessor getTssKeysAccessor() {
         return tssKeysAccessor;
     }
+
+    /**
+     * A class to manage the status change of the TSS.
+     * It computes the new status based on the old status and the current state of the system.
+     * If needed, it schedules work to generate TSS messages and votes.
+     */
+    private class StatusChange {
+        private TssKeyingStatus newKeyingStatus;
+        private RosterToKey newRosterToKey;
+        private Bytes newLedgerId;
+        private final boolean isStakePeriodBoundary;
+        private final Instant consensusNow;
+        private final RosterAndTssInfo info;
+
+        public StatusChange(
+                final boolean isStakePeriodBoundary, final Instant consensusNow, final RosterAndTssInfo info) {
+            this.isStakePeriodBoundary = isStakePeriodBoundary;
+            this.info = info;
+            this.newKeyingStatus = tssStatus.tssKeyingStatus();
+            this.newRosterToKey = tssStatus.rosterToKey();
+            this.newLedgerId = tssStatus.ledgerId();
+            this.consensusNow = consensusNow;
+        }
+
+        /**
+         * Computes the new status based on the old status and the current state of the system.
+         * If needed, it schedules work to generate TSS messages and votes.
+         *
+         * @return the new status
+         */
+        public TssStatus computeNewStatus() {
+            switch (tssStatus.rosterToKey()) {
+                case NONE -> {
+                    if (isStakePeriodBoundary) {
+                        newRosterToKey = RosterToKey.CANDIDATE_ROSTER;
+                        newKeyingStatus = TssKeyingStatus.WAITING_FOR_THRESHOLD_TSS_MESSAGES;
```

Comment:
Is it possible that isStakePeriodBoundary is true but we need to wait for more public tss encryption keys in state for the candidate roster? So perhaps this should be `WAITING_FOR_ENCRYPTION_KEYS`? Then it will transition to `WAITING_FOR_THRESHOLD_TSS_MESSAGES`

Reply:
If we have already keyed `ACTIVE_ROSTER` it means all encryption keys should be present right ?

Reply:
Yeah I think that's true, wanted to make sure.

Reply:
As discussed will add delay during bootstrap to get all encryption keys